# Data Collection

## Import packages

In [1]:
import pandas as pd
import numpy as np

## Import csv files

A .names file accompanied the csv files which provided descriptions for all attributes of the data including the original dtypes and feature descriptions.

In [2]:
assess = pd.read_csv('assessments.csv', na_values = '?', low_memory = False)
courses = pd.read_csv('courses.csv', na_values = '?', low_memory = False)
stassess = pd.read_csv('studentAssessment.csv', na_values = '?', low_memory = False)
stinfo = pd.read_csv('studentInfo.csv', na_values = '?', low_memory = False)
streg = pd.read_csv('studentRegistration.csv', na_values = '?', low_memory = False)
stvle = pd.read_csv('studentVle.csv', na_values = '?', low_memory = False)
vle = pd.read_csv('vle.csv', na_values = '?', low_memory = False)

# Preliminary data exploration

## Baseline Data Info

In [3]:
assess.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code_module        206 non-null    object 
 1   code_presentation  206 non-null    object 
 2   id_assessment      206 non-null    int64  
 3   assessment_type    206 non-null    object 
 4   date               195 non-null    float64
 5   weight             206 non-null    float64
dtypes: float64(2), int64(1), object(3)
memory usage: 9.8+ KB


In [4]:
assess.shape

(206, 6)

In [5]:
assess.head()

code_module code_presentation  id_assessment assessment_type   date  weight
0         AAA             2013J           1752             TMA   19.0    10.0
1         AAA             2013J           1753             TMA   54.0    20.0
2         AAA             2013J           1754             TMA  117.0    20.0
3         AAA             2013J           1755             TMA  166.0    20.0
4         AAA             2013J           1756             TMA  215.0    30.0

In [6]:
courses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 3 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   code_module                 22 non-null     object
 1   code_presentation           22 non-null     object
 2   module_presentation_length  22 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 660.0+ bytes


In [7]:
courses.shape

(22, 3)

In [8]:
courses.head()

code_module code_presentation  module_presentation_length
0         AAA             2013J                         268
1         AAA             2014J                         269
2         BBB             2013J                         268
3         BBB             2014J                         262
4         BBB             2013B                         240

In [9]:
stassess.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173912 entries, 0 to 173911
Data columns (total 5 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   id_assessment   173912 non-null  int64  
 1   id_student      173912 non-null  int64  
 2   date_submitted  173912 non-null  int64  
 3   is_banked       173912 non-null  int64  
 4   score           173739 non-null  float64
dtypes: float64(1), int64(4)
memory usage: 6.6 MB


In [10]:
stassess.shape

(173912, 5)

In [11]:
stassess.head()

id_assessment  id_student  date_submitted  is_banked  score
0           1752       11391              18          0   78.0
1           1752       28400              22          0   70.0
2           1752       31604              17          0   72.0
3           1752       32885              26          0   69.0
4           1752       38053              19          0   79.0

In [12]:
stinfo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   code_module           32593 non-null  object
 1   code_presentation     32593 non-null  object
 2   id_student            32593 non-null  int64 
 3   gender                32593 non-null  object
 4   region                32593 non-null  object
 5   highest_education     32593 non-null  object
 6   imd_band              31482 non-null  object
 7   age_band              32593 non-null  object
 8   num_of_prev_attempts  32593 non-null  int64 
 9   studied_credits       32593 non-null  int64 
 10  disability            32593 non-null  object
 11  final_result          32593 non-null  object
dtypes: int64(3), object(9)
memory usage: 3.0+ MB


In [13]:
stinfo.shape

(32593, 12)

In [14]:
stinfo.head()

code_module code_presentation  id_student gender                region  \
0         AAA             2013J       11391      M   East Anglian Region   
1         AAA             2013J       28400      F              Scotland   
2         AAA             2013J       30268      F  North Western Region   
3         AAA             2013J       31604      F     South East Region   
4         AAA             2013J       32885      F  West Midlands Region   

       highest_education imd_band age_band  num_of_prev_attempts  \
0       HE Qualification  90-100%     55<=                     0   
1       HE Qualification   20-30%    35-55                     0   
2  A Level or Equivalent   30-40%    35-55                     0   
3  A Level or Equivalent   50-60%    35-55                     0   
4     Lower Than A Level   50-60%     0-35                     0   

   studied_credits disability final_result  
0              240          N         Pass  
1               60          N         Pass  
2               60          Y    Withdrawn  
3               60          N         Pass  
4               60          N         Pass

In [15]:
streg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   code_module          32593 non-null  object 
 1   code_presentation    32593 non-null  object 
 2   id_student           32593 non-null  int64  
 3   date_registration    32548 non-null  float64
 4   date_unregistration  10072 non-null  float64
dtypes: float64(2), int64(1), object(2)
memory usage: 1.2+ MB


In [16]:
streg.shape

(32593, 5)

In [17]:
streg.head()

code_module code_presentation  id_student  date_registration  \
0         AAA             2013J       11391             -159.0   
1         AAA             2013J       28400              -53.0   
2         AAA             2013J       30268              -92.0   
3         AAA             2013J       31604              -52.0   
4         AAA             2013J       32885             -176.0   

   date_unregistration  
0                  NaN  
1                  NaN  
2                 12.0  
3                  NaN  
4                  NaN

In [18]:
stvle.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10655280 entries, 0 to 10655279
Data columns (total 6 columns):
 #   Column             Dtype 
---  ------             ----- 
 0   code_module        object
 1   code_presentation  object
 2   id_student         int64 
 3   id_site            int64 
 4   date               int64 
 5   sum_click          int64 
dtypes: int64(4), object(2)
memory usage: 487.8+ MB


In [19]:
stvle.shape

(10655280, 6)

In [20]:
stvle.head()

code_module code_presentation  id_student  id_site  date  sum_click
0         AAA             2013J       28400   546652   -10          4
1         AAA             2013J       28400   546652   -10          1
2         AAA             2013J       28400   546652   -10          1
3         AAA             2013J       28400   546614   -10         11
4         AAA             2013J       28400   546714   -10          1

In [21]:
vle.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6364 entries, 0 to 6363
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id_site            6364 non-null   int64  
 1   code_module        6364 non-null   object 
 2   code_presentation  6364 non-null   object 
 3   activity_type      6364 non-null   object 
 4   week_from          1121 non-null   float64
 5   week_to            1121 non-null   float64
dtypes: float64(2), int64(1), object(3)
memory usage: 298.4+ KB


In [22]:
vle.shape

(6364, 6)

In [23]:
vle.head()

id_site code_module code_presentation activity_type  week_from  week_to
0   546943         AAA             2013J      resource        NaN      NaN
1   546712         AAA             2013J     oucontent        NaN      NaN
2   546998         AAA             2013J      resource        NaN      NaN
3   546888         AAA             2013J           url        NaN      NaN
4   547035         AAA             2013J      resource        NaN      NaN

## Helper elements

I created helper elements that would be beneficial for some later looping code including a list of all the dataframes and assigning each df with a name

In [24]:
ls_all_docs = [assess, courses, stassess, stinfo, streg, stvle, vle]
assess.name = 'assess'
courses.name = 'courses'
stassess.name = 'stassess'
stinfo.name = 'stinfo'
streg.name = 'streg'
stvle.name = 'stvle'
vle.name = 'vle'

## Checking for Duplicates

In [25]:
for i in ls_all_docs:
    colu = [i.columns]
    for c in colu:
        if i[c].duplicated().any() is True:
            print(f'{i.name}[{c}] has duplicates')
        else:
            print(f'{i.name} has no duplicates')

assess has no duplicates
courses has no duplicates
stassess has no duplicates
stinfo has no duplicates
streg has no duplicates
stvle has no duplicates
vle has no duplicates


# Joining Data

## Finding merge features

I needed to identify the specific data features that were in common for each df. The first for loop required cross referencing multiple posts and sources (including my notes) to figure out the specific syntax.

In [26]:
similar_headers = []

for i, df_outer in enumerate(ls_all_docs):
    for j, df_inner in enumerate(ls_all_docs):
        if i == j:
            continue
        for c in df_outer.columns:
            if c in df_inner.columns and c not in similar_headers:
                similar_headers.append(c)

In [27]:
similar_headers

['code_module',
 'code_presentation',
 'id_assessment',
 'date',
 'id_student',
 'id_site']

In [28]:
for h in similar_headers:
    for d in ls_all_docs:
        ls = list(d.columns)
        if h in ls:
            print(f"{d.name} has {h}")

assess has code_module
courses has code_module
stinfo has code_module
streg has code_module
stvle has code_module
vle has code_module
assess has code_presentation
courses has code_presentation
stinfo has code_presentation
streg has code_presentation
stvle has code_presentation
vle has code_presentation
assess has id_assessment
stassess has id_assessment
assess has date
stvle has date
stassess has id_student
stinfo has id_student
streg has id_student
stvle has id_student
stvle has id_site
vle has id_site


With my common features identified, I then physically constructed a table to visualize which features overlap. To condense the number of repeat columns between merges, I chose to list all common features between data frames that were possible at a time.

## Joining attempt 1

All merges follow a similar pattern where all common features were used to merge onto. Additionally, df that were assumed to inform another df were merged first. The head method was added later to highlight something that I was not able to catch prior to merging.

Each merge was also checked to ensure there were no missing values. This will actually be identified as incorrect exemplified with 'sacri'.

as_co = assess.merge(courses, on = ['code_module', 'code_presentation'])
as_co.isna().sum()

as_co.info()

as_co.head()

as_co.shape

sta_ac = stassess.merge(as_co, on = 'id_assessment')
sta_ac.isna().sum()

sta_ac.info()

sta_ac.head()

sta_ac.shape

in_re = stinfo.merge(streg, on = ['code_module', 'code_presentation', 'id_student'])
in_re.isna().sum()

in_re.info()

in_re.head()

in_re.shape

sv_vl = stvle.merge(vle, on = ['id_site', 'code_module', 'code_presentation'])
sv_vl.isna().sum()

sv_vl.info()

sv_vl.head()

sv_vl.shape

sacri = sta_ac.merge(in_re, on = ['code_module', 'code_presentation', 'id_student'])
sacri.isna().sum()

sacri.info()

sacri.head()

sacri.shape

## Join error

The following merge produced error that indicated that the date feature in sacri is considered an object whereas the date column in sv_vl is considered an Int. Looking at the .names file accompanying document, the feature should be populated with integers indicating the time passed since the start of the course.

#merged_df = sv_vl.merge(sacri, on = ['code_module', 'code_presentation', 'date', 'id_student'])
#merged_df.isna().sum()

## Diagnosis

I first needed to identify what characters were being considered as objects rather than int's.

values_not_int = []
values_int = []

for index, row in sacri.iterrows():
    if row['date'] is not int:
        values_not_int.append(row['date'])
        print(f'{row["date"]}: {type(row["date"])}')
    else:
        values_int.append(row['date'])

values_int

The loop indicated that all values were considered strings. So I needed to convert all values to an integer.

The following code produced an error since '?' cannot be converted to int dtype.

#for index, row in sacri.iterrows():
#    row['date'] = int(row['date'])

sacri[sacri.eq('?').any(1)]

Seeing that date_unregistration had '?', I decided to check all other merge stages.

#in_re[in_re.eq('?').any(1)]

#sta_ac[sta_ac.eq('?').any(1)]

#as_co[as_co.eq('?').any(1)]

#sv_vl[sv_vl.eq('?').any(1)]

All merges have missing data in the form of '?' which is why the initial merges did not show any missing data.

## Missing original data

### Assess

Assess lists out the assignments associated with the respective modules and presentations.

assess[assess.eq('?').any(1)]

Since the missing data is the date the final submission date (int) for each assessment, we cannot drop those rows without damaging other df during merging. Since no supplemental data is available to determine the date, these dates will need to be imputed.

Assuming that the last day to turn in any assignment is the end of the course, each value was imputed with the maximum date within the feature.

assess['date'].replace('?', np.nan, inplace = True)
assess['date'] = pd.to_numeric(assess['date'], errors = 'coerce')
for c in assess.columns:
    assess_max = assess['date'].max()
    assess['date'].replace('?', assess_max, inplace = True)
#assess['date'] = assess['date'].astype('int64')

assess['date'].dtypes

assess.info()

assess.isnull().sum()

assess[assess['date'].isnull()]

assess.info()

courses[courses.eq('?').any(1)]

stassess[stassess.eq('?').any(1)]

stinfo[stinfo.eq('?').any(1)]

streg[streg.eq('?').any(1)]

stvle[stvle.eq('?').any(1)]

vle[vle.eq('?').any(1)]

print(sacri['date'].isna().sum()/len(sacri['date']))

I am choosing to dropna all rows that have had a '?' converted to NaN since the amount of missing data is less than 2%

sacri.dropna(subset = ['date'], inplace = True)
sacri.shape

## Joining attempt 2

After speaking with my mentor, I realized that there is a more streamline way to replace the '?' with NaN objects. I used the pd.read_csv() argument na_values = '?' to load the files with the NaN values. The following section will try to parse out what data I need to impute or manage prior to merging.

In [29]:
for i in ls_all_docs:
    print(i.name)
    print(i.isna().sum())
    print(i.shape)
    print()

assess
code_module           0
code_presentation     0
id_assessment         0
assessment_type       0
date                 11
weight                0
dtype: int64
(206, 6)

courses
code_module                   0
code_presentation             0
module_presentation_length    0
dtype: int64
(22, 3)

stassess
id_assessment       0
id_student          0
date_submitted      0
is_banked           0
score             173
dtype: int64
(173912, 5)

stinfo
code_module                0
code_presentation          0
id_student                 0
gender                     0
region                     0
highest_education          0
imd_band                1111
age_band                   0
num_of_prev_attempts       0
studied_credits            0
disability                 0
final_result               0
dtype: int64
(32593, 12)

streg
code_module                0
code_presentation          0
id_student                 0
date_registration         45
date_unregistration    22521
dtype: int64
(32593, 5)

### Assessment

Since the missing data is the date the final submission date (int) for each assessment, we cannot drop those rows without damaging other df during merging. Since no supplemental data is available to determine the date, these dates will need to be imputed.

Assuming that the last day to turn in any assignment is the end of the course, each value was imputed with the maximum date within the feature.

In [30]:
assess.shape

(206, 6)

In [31]:
assess['date'] = assess['date'].fillna(assess['date'].max())

In [32]:
assess.isna().sum()

code_module          0
code_presentation    0
id_assessment        0
assessment_type      0
date                 0
weight               0
dtype: int64

In [33]:
assess.shape

(206, 6)

### Student Assessment

In [34]:
stassess.shape

(173912, 5)

In [35]:
ranges = [0, 40, 50, 60, 70, 80, 90, 100]
stassess.groupby(pd.cut(stassess['score'], ranges)).count()

id_assessment  id_student  date_submitted  is_banked  score
score                                                                 
(0, 40]             9215        9215            9215       9215   9215
(40, 50]            7401        7401            7401       7401   7401
(50, 60]           16561       16561           16561      16561  16561
(60, 70]           23920       23920           23920      23920  23920
(70, 80]           40752       40752           40752      40752  40752
(80, 90]           36330       36330           36330      36330  36330
(90, 100]          39231       39231           39231      39231  39231

Missing values imply that the grade was not able to be reported for each of the 173 student assignments. Since the scores are being investigated as dependent variables, their presence is crucial to the study. Dropping the 173 student assignments will minimally (0.09% of the sample population) impact the study.

In [36]:
stassess_drop = stassess.dropna()

In [37]:
stassess_drop.isna().sum()

id_assessment     0
id_student        0
date_submitted    0
is_banked         0
score             0
dtype: int64

In [38]:
stassess_drop.shape

(173739, 5)

### Student Info

In [39]:
stinfo.shape

(32593, 12)

In [40]:
stinfo.head()

code_module code_presentation  id_student gender                region  \
0         AAA             2013J       11391      M   East Anglian Region   
1         AAA             2013J       28400      F              Scotland   
2         AAA             2013J       30268      F  North Western Region   
3         AAA             2013J       31604      F     South East Region   
4         AAA             2013J       32885      F  West Midlands Region   

       highest_education imd_band age_band  num_of_prev_attempts  \
0       HE Qualification  90-100%     55<=                     0   
1       HE Qualification   20-30%    35-55                     0   
2  A Level or Equivalent   30-40%    35-55                     0   
3  A Level or Equivalent   50-60%    35-55                     0   
4     Lower Than A Level   50-60%     0-35                     0   

   studied_credits disability final_result  
0              240          N         Pass  
1               60          N         Pass  
2               60          Y    Withdrawn  
3               60          N         Pass  
4               60          N         Pass

In [41]:
stinfo['imd_band'].value_counts()

20-30%     3654
30-40%     3539
10-20      3516
0-10%      3311
40-50%     3256
50-60%     3124
60-70%     2905
70-80%     2879
80-90%     2762
90-100%    2536
Name: imd_band, dtype: int64

In [42]:
stinfo.loc[(stinfo['imd_band'] == '10-20'),'imd_band'] = stinfo['imd_band'] + '%'
stinfo['imd_band'].value_counts()

20-30%     3654
30-40%     3539
10-20%     3516
0-10%      3311
40-50%     3256
50-60%     3124
60-70%     2905
70-80%     2879
80-90%     2762
90-100%    2536
Name: imd_band, dtype: int64

Since 3% of the data would be impacted, the rows with missing data will be dropped. It is unknown if the economic status (imd_band) would impact the interactions so we cannot drop that column.

In [43]:
stinfo_drop = stinfo.dropna()
stinfo_drop.isna().sum()

code_module             0
code_presentation       0
id_student              0
gender                  0
region                  0
highest_education       0
imd_band                0
age_band                0
num_of_prev_attempts    0
studied_credits         0
disability              0
final_result            0
dtype: int64

In [44]:
stinfo_drop.shape

(31482, 12)

### Student Registration

In [45]:
streg.shape

(32593, 5)

In [46]:
streg.head()

code_module code_presentation  id_student  date_registration  \
0         AAA             2013J       11391             -159.0   
1         AAA             2013J       28400              -53.0   
2         AAA             2013J       30268              -92.0   
3         AAA             2013J       31604              -52.0   
4         AAA             2013J       32885             -176.0   

   date_unregistration  
0                  NaN  
1                  NaN  
2                 12.0  
3                  NaN  
4                  NaN

The goal is to determine features that may predict student success and engagement for these reasons, the students that unregistered will be dropped and the date of unregistration will be ignored (drop).

In [47]:
streg_completed = streg[pd.isnull(streg['date_unregistration'])]
streg_completed.shape

(22521, 5)

In [48]:
streg_drop = streg_completed.drop(columns = 'date_unregistration')
streg_drop.shape

(22521, 4)

In [49]:
streg_drop.isna().sum()

code_module          0
code_presentation    0
id_student           0
date_registration    6
dtype: int64

In [50]:
streg_df = streg_drop.dropna()
streg_df.isna().sum()

code_module          0
code_presentation    0
id_student           0
date_registration    0
dtype: int64

In [51]:
streg_df.shape

(22515, 4)

### VLE

The start and end dates of the modules are going to be considered not relavent to whether students succeed or fail. The two columns (week_from and week_to) will be dropped from analysis.

In [52]:
vle.shape

(6364, 6)

In [53]:
vle.head()

id_site code_module code_presentation activity_type  week_from  week_to
0   546943         AAA             2013J      resource        NaN      NaN
1   546712         AAA             2013J     oucontent        NaN      NaN
2   546998         AAA             2013J      resource        NaN      NaN
3   546888         AAA             2013J           url        NaN      NaN
4   547035         AAA             2013J      resource        NaN      NaN

In [54]:
vle_drop = vle.drop(columns = ['week_from', 'week_to'])
vle_drop.shape

(6364, 4)

In [55]:
as_co = assess.merge(courses, on = ['code_module', 'code_presentation'])
as_co.isna().sum()

code_module                   0
code_presentation             0
id_assessment                 0
assessment_type               0
date                          0
weight                        0
module_presentation_length    0
dtype: int64

In [56]:
as_co.shape

(206, 7)

In [57]:
as_co.head()

code_module code_presentation  id_assessment assessment_type   date  weight  \
0         AAA             2013J           1752             TMA   19.0    10.0   
1         AAA             2013J           1753             TMA   54.0    20.0   
2         AAA             2013J           1754             TMA  117.0    20.0   
3         AAA             2013J           1755             TMA  166.0    20.0   
4         AAA             2013J           1756             TMA  215.0    30.0   

   module_presentation_length  
0                         268  
1                         268  
2                         268  
3                         268  
4                         268

In [58]:
sta_ac = stassess_drop.merge(as_co, on = 'id_assessment')
sta_ac.isna().sum()

id_assessment                 0
id_student                    0
date_submitted                0
is_banked                     0
score                         0
code_module                   0
code_presentation             0
assessment_type               0
date                          0
weight                        0
module_presentation_length    0
dtype: int64

In [59]:
sta_ac.head()

id_assessment  id_student  date_submitted  is_banked  score code_module  \
0           1752       11391              18          0   78.0         AAA   
1           1752       28400              22          0   70.0         AAA   
2           1752       31604              17          0   72.0         AAA   
3           1752       32885              26          0   69.0         AAA   
4           1752       38053              19          0   79.0         AAA   

  code_presentation assessment_type  date  weight  module_presentation_length  
0             2013J             TMA  19.0    10.0                         268  
1             2013J             TMA  19.0    10.0                         268  
2             2013J             TMA  19.0    10.0                         268  
3             2013J             TMA  19.0    10.0                         268  
4             2013J             TMA  19.0    10.0                         268

In [60]:
sta_ac.shape

(173739, 11)

In [61]:
in_re = stinfo_drop.merge(streg_df, on = ['code_module', 'code_presentation', 'id_student'])
in_re.isna().sum()

code_module             0
code_presentation       0
id_student              0
gender                  0
region                  0
highest_education       0
imd_band                0
age_band                0
num_of_prev_attempts    0
studied_credits         0
disability              0
final_result            0
date_registration       0
dtype: int64

In [62]:
in_re.head()

code_module code_presentation  id_student gender                region  \
0         AAA             2013J       11391      M   East Anglian Region   
1         AAA             2013J       28400      F              Scotland   
2         AAA             2013J       31604      F     South East Region   
3         AAA             2013J       32885      F  West Midlands Region   
4         AAA             2013J       38053      M                 Wales   

       highest_education imd_band age_band  num_of_prev_attempts  \
0       HE Qualification  90-100%     55<=                     0   
1       HE Qualification   20-30%    35-55                     0   
2  A Level or Equivalent   50-60%    35-55                     0   
3     Lower Than A Level   50-60%     0-35                     0   
4  A Level or Equivalent   80-90%    35-55                     0   

   studied_credits disability final_result  date_registration  
0              240          N         Pass             -159.0  
1               60          N         Pass              -53.0  
2               60          N         Pass              -52.0  
3               60          N         Pass             -176.0  
4               60          N         Pass             -110.0

In [63]:
in_re.shape

(21639, 13)

In [64]:
sv_vl = stvle.merge(vle_drop, on = ['id_site', 'code_module', 'code_presentation'])
sv_vl.isna().sum()

code_module          0
code_presentation    0
id_student           0
id_site              0
date                 0
sum_click            0
activity_type        0
dtype: int64

In [65]:
sv_vl.head()

code_module code_presentation  id_student  id_site  date  sum_click  \
0         AAA             2013J       28400   546652   -10          4   
1         AAA             2013J       28400   546652   -10          1   
2         AAA             2013J       28400   546652   -10          1   
3         AAA             2013J       28400   546652   -10          8   
4         AAA             2013J       30268   546652   -10          3   

  activity_type  
0       forumng  
1       forumng  
2       forumng  
3       forumng  
4       forumng

In [66]:
sv_vl.shape

(10655280, 7)

In [67]:
sacri = sta_ac.merge(in_re, on = ['code_module', 'code_presentation', 'id_student'])
sacri.isna().sum()

id_assessment                 0
id_student                    0
date_submitted                0
is_banked                     0
score                         0
code_module                   0
code_presentation             0
assessment_type               0
date                          0
weight                        0
module_presentation_length    0
gender                        0
region                        0
highest_education             0
imd_band                      0
age_band                      0
num_of_prev_attempts          0
studied_credits               0
disability                    0
final_result                  0
date_registration             0
dtype: int64

In [68]:
sacri.head()

id_assessment  id_student  date_submitted  is_banked  score code_module  \
0           1752       11391              18          0   78.0         AAA   
1           1753       11391              53          0   85.0         AAA   
2           1754       11391             115          0   80.0         AAA   
3           1755       11391             164          0   85.0         AAA   
4           1756       11391             212          0   82.0         AAA   

  code_presentation assessment_type   date  weight  ...  gender  \
0             2013J             TMA   19.0    10.0  ...       M   
1             2013J             TMA   54.0    20.0  ...       M   
2             2013J             TMA  117.0    20.0  ...       M   
3             2013J             TMA  166.0    20.0  ...       M   
4             2013J             TMA  215.0    30.0  ...       M   

                region highest_education imd_band age_band  \
0  East Anglian Region  HE Qualification  90-100%     55<=   
1  East Anglian Region  HE Qualification  90-100%     55<=   
2  East Anglian Region  HE Qualification  90-100%     55<=   
3  East Anglian Region  HE Qualification  90-100%     55<=   
4  East Anglian Region  HE Qualification  90-100%     55<=   

  num_of_prev_attempts  studied_credits  disability final_result  \
0                    0              240           N         Pass   
1                    0              240           N         Pass   
2                    0              240           N         Pass   
3                    0              240           N         Pass   
4                    0              240           N         Pass   

  date_registration  
0            -159.0  
1            -159.0  
2            -159.0  
3            -159.0  
4            -159.0  

[5 rows x 21 columns]

In [69]:
sacri.shape

(153476, 21)

In [70]:
df = sv_vl.merge(sacri, on = ['code_module', 'code_presentation', 'date', 'id_student'])
df.isna().sum()

code_module                   0
code_presentation             0
id_student                    0
id_site                       0
date                          0
sum_click                     0
activity_type                 0
id_assessment                 0
date_submitted                0
is_banked                     0
score                         0
assessment_type               0
weight                        0
module_presentation_length    0
gender                        0
region                        0
highest_education             0
imd_band                      0
age_band                      0
num_of_prev_attempts          0
studied_credits               0
disability                    0
final_result                  0
date_registration             0
dtype: int64

In [71]:
df.head()

code_module code_presentation  id_student  id_site  date  sum_click  \
0         AAA             2013J     2412002   546652    19          1   
1         AAA             2013J     2412002   546652    19          1   
2         AAA             2013J     2412002   546614    19         10   
3         AAA             2013J     2412002   546876    19          2   
4         AAA             2013J     2412002   546871    19          1   

  activity_type  id_assessment  date_submitted  is_banked  ...  gender  \
0       forumng           1752              20          0  ...       M   
1       forumng           1752              20          0  ...       M   
2      homepage           1752              20          0  ...       M   
3       subpage           1752              20          0  ...       M   
4       subpage           1752              20          0  ...       M   

              region   highest_education  imd_band age_band  \
0  South West Region  Lower Than A Level    50-60%    35-55   
1  South West Region  Lower Than A Level    50-60%    35-55   
2  South West Region  Lower Than A Level    50-60%    35-55   
3  South West Region  Lower Than A Level    50-60%    35-55   
4  South West Region  Lower Than A Level    50-60%    35-55   

  num_of_prev_attempts studied_credits disability final_result  \
0                    0              90          N  Distinction   
1                    0              90          N  Distinction   
2                    0              90          N  Distinction   
3                    0              90          N  Distinction   
4                    0              90          N  Distinction   

   date_registration  
0              -38.0  
1              -38.0  
2              -38.0  
3              -38.0  
4              -38.0  

[5 rows x 24 columns]

In [72]:
df.shape

(534781, 24)

In [73]:
col_order = ['id_student','gender', 'disability', 'age_band', 
             'region', 'imd_band', 'highest_education', 'id_site', 
             'code_module', 'code_presentation', 'id_assessment', 
             'date_registration', 'num_of_prev_attempts', 
             'final_result', 'studied_credits', 'date_submitted', 
             'score', 'assessment_type', 'module_presentation_length', 
             'weight', 'is_banked',  'activity_type', 'date', 'sum_click']
df = df[col_order]
df.head()

id_student gender disability age_band             region imd_band  \
0     2412002      M          N    35-55  South West Region   50-60%   
1     2412002      M          N    35-55  South West Region   50-60%   
2     2412002      M          N    35-55  South West Region   50-60%   
3     2412002      M          N    35-55  South West Region   50-60%   
4     2412002      M          N    35-55  South West Region   50-60%   

    highest_education  id_site code_module code_presentation  ...  \
0  Lower Than A Level   546652         AAA             2013J  ...   
1  Lower Than A Level   546652         AAA             2013J  ...   
2  Lower Than A Level   546614         AAA             2013J  ...   
3  Lower Than A Level   546876         AAA             2013J  ...   
4  Lower Than A Level   546871         AAA             2013J  ...   

   studied_credits  date_submitted  score assessment_type  \
0               90              20   86.0             TMA   
1               90              20   86.0             TMA   
2               90              20   86.0             TMA   
3               90              20   86.0             TMA   
4               90              20   86.0             TMA   

   module_presentation_length  weight  is_banked activity_type  date  \
0                         268    10.0          0       forumng    19   
1                         268    10.0          0       forumng    19   
2                         268    10.0          0      homepage    19   
3                         268    10.0          0       subpage    19   
4                         268    10.0          0       subpage    19   

   sum_click  
0          1  
1          1  
2         10  
3          2  
4          1  

[5 rows x 24 columns]

In [74]:
typo_check = ['gender', 'disability', 'age_band', 
             'region', 'imd_band', 'highest_education', 
             'date_registration', 'num_of_prev_attempts', 
             'final_result', 'studied_credits', 'date_submitted', 
             'score', 'assessment_type', 'module_presentation_length', 
             'weight', 'is_banked',  'activity_type', 'date', 'sum_click']

for c in typo_check:
    print(c)
    print(df[c].unique())
    print()

gender
['M' 'F']

disability
['N' 'Y']

age_band
['35-55' '0-35' '55<=']

region
['South West Region' 'East Midlands Region' 'East Anglian Region'
 'Scotland' 'West Midlands Region' 'Yorkshire Region' 'Ireland'
 'South Region' 'North Western Region' 'North Region' 'London Region'
 'Wales' 'South East Region']

imd_band
['50-60%' '10-20%' '40-50%' '60-70%' '90-100%' '30-40%' '70-80%' '20-30%'
 '0-10%' '80-90%']

highest_education
['Lower Than A Level' 'A Level or Equivalent' 'HE Qualification'
 'Post Graduate Qualification' 'No Formal quals']

date_registration
[ -38. -127.  -71.  -58. -144.  -80. -191.    2.  -68.  -49.  -67.  -18.
  -50. -195.  -32.  -54.  -15.  -16.  -43.  -73.  -46. -158.  -52.  -11.
  -81.  -87.  -48.  -31.  -94. -109.  -35.  -30.  -63.  -74.  -29. -100.
 -165. -128.  -25. -101.  -91. -138.  -57. -108.  -62. -102. -164.  -53.
 -106.  -56.  -85.  -24.  -22. -185. -137. -148.  -45. -123. -113.  -66.
  -59.  -95. -120.  -61. -129. -130.  -23. -156.   -4. -155. -172. -